File: c/knight.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

#define BOARD_N 8 /* length of line */
#define BOARD_M 8 /* no of lines */
#define NO_OF_NEIGHBOURS 8

typedef int neighbours_t[NO_OF_NEIGHBOURS];

void init_board(int board[])
{
	int i;
	for (i=0; i<BOARD_N*BOARD_M; i++) {
		board[i] = -1;
	}
}

int board_completed(int board[])
{
	int i;
	for (i=0; i<BOARD_N*BOARD_M; i++) {
		if (board[i] == -1)
			return -1;
	}
	return 0;
}

void move_from_to(int board[], int from_pos, int to_pos)
{
	board[from_pos] = to_pos;
}

int board_is_empty_pos(int board[], int pos)
{
	if (board[pos] == -1)
		return 0;

	return -1;
}

int board_is_valid_pos(int board[], int pos)
{
	if (pos < 0 || pos >= BOARD_N*BOARD_M)
		return -1;

	return 0;
}

int board_get_line(int pos)
{
	return pos / BOARD_N;
}

int get_free_neighbour(int board[], int pos, neighbours_t neighbours, int i, int dx, int dy)
{
	int neighbour;

	if (neighbours != NULL) {
		neighbours[i] = -1;
	}
	neighbour = pos + (dx) + (dy) * 8; // <dx,dy>
	if (board_is_valid_pos(board, neighbour) == 0 && board_get_line(pos) + dy == board_get_line(neighbour) && board_is_empty_pos(board, neighbour) == 0) {
		if (neighbours != NULL) {
			neighbours[i] = neighbour;
		}
		return 0;
	}
	return -1;
}

int get_free_neighbours(int board[], int pos, neighbours_t neighbours)
{
	int no_of_neighbours = 0;

	if (get_free_neighbour(board, pos, neighbours, 0, -1, -2) == 0) //1 <-1, -2>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 1, +1, -2) == 0) //2 <+1, -2>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 2, +2, -1) == 0) //3 <+2, -1>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 3, +2, +1) == 0) //4 <+2, +1>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 4, +1, +2) == 0) //5 <+1, +2>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 5, -1, +2) == 0) //6 <-1, +2>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 6, -2, +1) == 0) //7 <-2, +1>
		no_of_neighbours++;

	if (get_free_neighbour(board, pos, neighbours, 7, -2, -1) == 0) //8 <-2, -1>
		no_of_neighbours++;

	return no_of_neighbours;
}

int choose_best_neighbour(int board[], neighbours_t neighbours)
{
	int i;
	int best_neighbour_neighbours = BOARD_N * BOARD_M; //Unreachable high value
	int best_neighbour = -1;

	for (i=0; i<NO_OF_NEIGHBOURS; i++) {
		if (neighbours[i] != -1 && get_free_neighbours(board, neighbours[i], NULL) < best_neighbour_neighbours) {
			best_neighbour = neighbours[i];
			best_neighbour_neighbours = get_free_neighbours(board, neighbours[i], NULL);
		}
	}
	return best_neighbour;
}

int solve_from_pos(int pos)
{
	int board[BOARD_M * BOARD_N];
	neighbours_t neighbours;

	init_board(board);

	while (get_free_neighbours(board, pos, neighbours) > 0) {
		int best_neighbour = choose_best_neighbour(board, neighbours);
		move_from_to(board, pos, best_neighbour);
		pos = best_neighbour;
	}

	move_from_to(board, pos, pos);

	if (board_completed(board) != 0) {
		return -1;
	}

	return 0;
}

int main()
{
	int pos;
	int ret;
	int good_start_pos;

	good_start_pos = 0;
	for (pos=0; pos < BOARD_M*BOARD_N; pos++) {
		ret = solve_from_pos(pos);
		if (ret == 0) {
			printf("A solution for start position %d found.\n", pos);
			good_start_pos++;
		} else {
			printf("No solution for start position %d found.\n", pos);
		}
	}

	printf("Found a solution in %d out of %d cases.\n", good_start_pos, BOARD_M*BOARD_N);

	return 0;
}

File: gforth/knight.fs
\ Knight's problem solver
\
\ The following parameters are adjustable
\ n ... length of board
\ m ... width of board
\ neighbour_precendence ... precedence which defines which neighbour
\   will be choosen if, there is more than one posibility

\ Here are the numbers of the possible neighbours:
\ +-+-+-+-+-+
\ | |1| |2| |
\ +-+-+-+-+-+
\ |8| | | |3|
\ +-+-+-+-+-+
\ | | |*| | |
\ +-+-+-+-+-+
\ |7| | | |4|
\ +-+-+-+-+-+
\ | |6| |5| |
\ +-+-+-+-+-+

8 constant n
8 constant m
create neighbour_precedence 5 , 4 , 2 , 6 , 8 , 3 , 1 , 7 ,

create board n m * cells allot
create neighbours 8 cells allot
create possible_neighbours_dx -1 , 1 , 2 , 2 , 1 , -1 , -2 , -2 ,
create possible_neighbours_dy -2 , -2 , -1 , 1 , 2 , 2 , 1 , -1 ,

: init_board ( -- )
	n m * 0 do -1 board I cells + ! loop ;

: board_move_from_to ( from_pos to_pos -- )
 	swap board swap cells + ! ;

: board_is_empty_pos ( pos -- t/f )
	board swap cells + @ -1 = ;

: board_completed ( -- t/f )
	n m * 0 do
		i board_is_empty_pos
			if
				false unloop exit
			endif
	loop
	true ;

: board_is_valid_pos ( pos -- t/f )
	dup 0 >=
	swap n m * <
	and ;

: board_get_line ( pos -- t/f )
	n / ;

: calc_neighbour_pos ( neighbour_id -- dx dy )
	dup
	possible_neighbours_dx swap cells + @
	swap
	possible_neighbours_dy swap cells + @ ;

: get_free_neighbour ( pos dx dy - neighbour t/f )
	rot ( dx dy pos )
	2dup swap >r >r ( dx dy pos )
	swap n ( dx pos dy n )
	* + + ( neighbour )
	dup board_is_valid_pos invert if r> r> 2drop false exit endif
	dup board_get_line r> board_get_line r> + = invert if false exit endif
	dup board_is_empty_pos invert if false exit endif
	true ;

: get_free_neighbours_raw ( pos -- no_of_neighbours )
	0 \ no_of_neighbours

	8 0 do
		over i calc_neighbour_pos get_free_neighbour nip
		if 1 + endif
	loop
	nip ;

: get_free_neighbours ( pos -- no_of_neighbours )
	0 \ no_of_neighbours
	8 0 do -1 neighbours i cells + ! loop

	8 0 do
		over i calc_neighbour_pos get_free_neighbour
		if
			neighbours i cells + !
			1 +
		else
			drop
		endif
	loop
	nip ;

: choose_best_neighbour ( -- best_neighbour )
	-1 ( best_neighbour = invalid position)
	n m * ( best_neighbour_neighbours = unreachable high value )

	8 0 do
		neighbour_precedence i cells + @ 1 - ( bn bnn n )
		neighbours swap cells + @ ( bn bnn neighbour )
		dup -1 <> if
			-rot 2>r ( neighbour | bn bnn )
			dup get_free_neighbours_raw ( neighbour #n | bn bnn )
			dup r@ ( neighbour #n bnn | bn bnn )
			< if ( neighbour #n | bn bnn)
				2rdrop ( neighbour #n)
			else
				2drop 2r> ( bn bnn )
			endif
		else
			drop
		endif
	loop
	drop ;

: solve_from_pos ( pos -- success )
	init_board

	n m * 0 do
		dup						( pos pos )
		get_free_neighbours 0 = ( pos t/f )
		if leave endif			( pos )
		choose_best_neighbour	( pos best_neighbour )
		2dup					( pos best_neighbour pos best_neighbour)
		board_move_from_to		( pos best_neighbour )
		nip 					( best_neighbour )
	loop

	dup
	board_move_from_to
	
	board_completed ;

: display_row ( row -- )
    dup
    48 + emit ( print line no )
    n 0 do
	124 emit ( | )
	dup board swap m * i + cells + @ .
    loop
    124 emit
    cr
	drop ;

: display
    cr
    m 0 do
	i display_row
	loop
	cr cr ;

: solve_all
	0 ( number of positions where we found a solution )
	n m * 0 do
		dup i solve_from_pos
		if
			." Found a solution for startpoint " . cr
			1 +
		else
			." Didn't find a solution for startpoint " . cr
		endif
		display
	loop

	." Found a solution in " . ." out of " n m * . ." cases." cr ;

: solve_one ( pos -- )
	dup solve_from_pos
	if
		cr ." Found a solution for startpoint " . cr
	else
		cr ." Didn't find a solution for startpoint " . cr
	endif ;

: main
	solve_all
	\ 2 solve_one
	;

main
bye
